### PR TITLE
fix(subagent): stop orphaned subagents at parent session end via SESSION_END hook

### LINKED ISSUE #3259: fix(subagent): stop still-running subagents at parent session end (production teardown)
Slice 3 of the #3191 control plane. Full design: https://github.com/gptme/gptme/issues/3191#issuecomment-4982523398

## Problem

Nothing in production stops still-running subagents when the parent conversation/session ends — only the test fixture does (`tests/conftest.py:328-374`). In long-lived hosts (gptme-server, TUI, an interactive chat that exits mid-fan-out), subprocess-mode subagents are orphaned: they keep running and burning tokens after their orchestrator is gone, with nobody left to read the result.

## Proposal

A `SESSION_END` hook registered by the subagent tool (the tool already registers a LOOP_CONTINUE hook, `subagent/__init__.py:530-536`):

- Snapshot `_subagents` under `_subagents_lock` (copy-before-iterate, same as `subagent_list`).
- For each agent without a terminal result: call `subagent_cancel(agent_id)` — today that terminates subprocesses and cache-marks threads; once #3258 lands, threads stop cooperatively too, with no change needed here.
- Grace period then `kill()` for subprocesses that ignore terminate (the cancel path already does 5s escalation, `api.py:1004-1009`).
- Log a one-line summary (`stopped N running subagents at session end`).

Scope note: thread-mode agents are daemon threads that die with the process in CLI runs — the real target is subprocess orphans and long-lived server/TUI processes.

## Acceptance criteria

- Parent `chat()` exit with a running subprocess subagent → child process is reaped (no orphan `python -m gptme` left).
- Terminal results already cached are not overwritten (existing `set_subagent_result_if_absent` semantics).
- No teardown hang: bounded joins/waits (test fixture uses 2s; production can afford ~5s).
- Test covering the SESSION_END path in both modes.

### PR DESCRIPTION
**Pull request by TimeToBuildBob**
Adds a `SESSION_END` hook to the subagent tool that cancels any still-running subagents without a terminal cached result when the parent conversation closes. This prevents leaked subprocess children and zombie thread-mode subagents with dangling concurrency slots.

**Implementation:**
- New `_session_end_subagent_cleanup` hook function (`gptme/tools/subagent/hooks.py`)
- Registered in ToolSpec as `("session.end", ...)` so it loads only when subagent tool is active
- Snapshots `_subagents` under `_subagents_lock`, skips agents with cached terminal results or dead threads, cancels the rest
- Subprocess cancel: SIGTERM → SIGKILL after 5s (bounded, handled by `subagent_cancel`)
- Thread cancel: result pre-marked as cancelled (non-blocking), thread stops at next checkpoint
- Handles all three edge cases: running (cancelled), terminated-with-result (skipped), not-running (skipped)

**Test:**
- `test_session_end_teardown_cancels_running_subagents` — verifies a running subagent without cached result is cancelled, and that terminated/dead subagents are skipped
- 122 existing subagent tests continue to pass

Closes #3259

### CODE REVIEW

**Comment by TimeToBuildBob** [comment]
@greptileai review

File: gptme/tools/subagent/hooks.py [hunk 1]
```diff
@@ -147,6 +160,50 @@ def notify_completion(agent_id: str, status: Status, summary: str) -> None:
...
+    waiting for them, but the Python thread continues to its next natural
+    checkpoint and then stops.
+
+    Bounded: the per-agent cancel wait is capped at ~5s (subprocess
+    SIGKILL escalation); the overall hook never blocks indefinitely.
+    """
+    from .api import subagent_cancel  # avoid circular import
+
+    with _subagents_lock:
+        snapshot = list(_subagents)
```
[the remark this answers is not in the document]
Comment (TimeToBuildBob) [reply gptme/tools/subagent/hooks.py:195]: Fixed in 8ddfcf605. Added `parent_logdir: Path | None` field to `Subagent` and now capture it unconditionally from `LogManager.get_current_log()` at spawn time. The SESSION_END hook now filters the snapshot to only cancel agents whose `parent_logdir == manager.logdir`; subagents owned by other conversations are skipped. Test updated with an explicit cross-session isolation case ("test-other-session" with a different parent_logdir is asserted NOT cancelled).

File: gptme/tools/subagent/hooks.py [hunk 2]
```diff
@@ -147,6 +160,50 @@ def notify_completion(agent_id: str, status: Status, summary: str) -> None:
...
+        with _subagent_results_lock:
+            has_terminal = sa.agent_id in _subagent_results
+
+        if has_terminal:
+            continue
+        if not sa.is_running():
+            continue
+
+        try:
+            subagent_cancel(sa.agent_id)
```
[the remark this answers is not in the document]
Comment (TimeToBuildBob) [reply gptme/tools/subagent/hooks.py:213]: Acknowledged. Python threads cannot be forcibly stopped — this is a known language-level limitation. The docstring has been updated to be explicit: 'the underlying Python thread is not forcibly stopped (Python does not support forcible thread termination) — it continues until its next natural checkpoint checks the cached result or the process exits.' The cancel-and-mark path still prevents the orchestrator from blocking and marks the concurrency slot released. Full forcible-stop support would require a coordinated threading.Event checked at every subagent tool call — a larger followup change beyond this PR's scope.

File: gptme/tools/subagent/api.py [hunk 3]
```diff
@@ -400,6 +405,7 @@ def subagent(
             redact_secrets=redact_secrets,
             context_window=context_window,
             max_time=max_time,
+            parent_logdir=parent_logdir,
```
[the remark this answers is not in the document]
Comment (TimeToBuildBob) [reply gptme/tools/subagent/api.py:408]: Fixed in 5dd123d37. Two changes together close the gap:
1. `subagent()` in api.py now passes `parent_logdir=parent_logdir` to the `_run_planner()` call (it was missing from that call path).
2. `_run_planner()` in execution.py now accepts `parent_logdir` as a parameter and propagates it to all three executor Subagent constructor calls (subprocess mode, thread-cleanup mode, and thread-final mode).

Executor subagents now have `parent_logdir=<session_logdir>`, so the SESSION_END hook's ownership filter correctly cancels them when their parent session ends.

Also added `test_session_end_teardown_cancels_planner_executors` test to cover this path explicitly.

**Comment by TimeToBuildBob** [comment]
@greptileai review
